twisted/internet/defer.py

```diff
@@ -661,6 +661,66 @@ def __str__(self):
     __repr__ = __str__
 
 
+    def __await__(self):
+        awaiter = _MakeDeferredAwaitable(self)
+        return awaiter.__await__()
+
+
+
+class _MakeDeferredAwaitable(object):
+    """
+    A subgenerator which is created by L{Deferred.__await__}.
+    """
+    result = _NO_RESULT
+
+    def __init__(self, d):
+        self.d = d
+        self.d.addBoth(self._setResult)
+
+    def _setResult(self, value):
+        self.result = value
+
+    def __next__(self):
+        if self.result is not _NO_RESULT:
+            raise StopIteration(self.result)
+        return self.d
+
+    def __await__(self):
```

Comment:
Actually it should only implement `__iter__`, and no `__await__`:

```python
class Deferred:
    def __await__(self):
        return _MakeDeferredAwaitable(self)

class _MakeDeferredAwaitable(object):
    def __iter__(self):
        return self

    def __next__(self):
         if self.result is not _NO_RESULT:
             raise StopIteration(self.result)
         return self.d
```

Per PEP 492 -- `__await__` should return an iterator (not an awaitable).  Also, you might want to implement `_MakeDeferredAwaitable.throw` if it's possible in Twisted to throw exceptions to the `await` points.